### DISCUSSION TITLE
Preferences: Livewire component(s) with table and CRUD modal form

### DISCUSSION THREAD

**Discussion by andrewmarx** [livewire/livewire#3695] [4. General]
I've recently started learning livewire and alpine (after picking up php and Laravel a month ago) and I'm struggling to tie some of the concepts together into the bigger picture.

I have a blade document, and in it I use a Livewire component to display the contents of a table. Simple enough. My question comes with the modal form to edit records; each row in the table has an edit button, and when the user hits it, it opens the modal and populates the form fields with the current data.

it seems like I have a few options with it:

1. I can create an entirely separate Livewire component for the modal form. When I'm trying to search for tutorials on modal forms with Livewire, this seems to be what I find, and I can get it to work.
2. I can use just a single component for both the table and modal and have functions in the component related to both the table and the CRUD operations. Then I can use Livewire to help populate the form fields when the user hits an 'edit' button to display the modal.
3.  Same as the previous option, but decouple the modal code from Livewire so that the only interaction with Livewire occurs when the user hits submit. This requires some other approach to populate fields in the form when the user hits an 'edit' button to bring up the modal

Option 1 has been the easiest because of the variety of tutorials on the topic, but the tutorials aren't generally discussing the topic in the context of displaying table data. So I'm not sure if there are any good practices for one vs two components when needing to do both.

Option 2 is doable for me, and I like having just a single component to handle everything related to the table and form. But populating the modal fields requires a round-trip request to the server, which seems unnecessary.

Option 3 is what I've been trying to do; I like the idea of opening the modal and instantly populating the fields from the client without bugging the server, then only communicate with the server when the form is submitted. Due to my inexperience with web-programming, I'm still trying to get this to work. My attempts all eventually turn into a convoluted mess trying to pass the data to the modal to populate the fields. But maybe it doesn't matter (this is more of an internal use app with a small user base) and I should just go with one of the other options for now.

Some I'm curious how others would approach this type of problem. One or two components? Couple or decouple the modal from Livewire?

**Comment by jonneroelofs** [comment]
If you sponsor Livewire you can watch screencasts by the author (of Livewire and Alpine). He has a series that will answer your questions and more.


![image](https://user-images.githubusercontent.com/20800818/132949452-2004fe45-c30f-42d2-9670-94a63bc747cf.png)

**Comment by jonneroelofs** [comment]
And to actually answer your question:
- I have an Abstract Datatable Livewire component, that I extend any time I need a Datatable.
- I have a dedicated blade/alpine component for a generic modal
- I have dedicated Livewire components for Crud that re-use the generic blade/alpine modal.
- The Datatable component is responsible for rendering the dedicated crud livewire components and injecting the attributes when a user clicks a row edit button for example.
